gcc/rust/ast/rust-expr.h

```diff
@@ -139,6 +139,56 @@ class AttrInputLiteral : public AttrInput
   }
 };
 
+// Like an AttrInputLiteral, but stores a MacroInvocation
+class AttrInputMacro : public AttrInput
+{
+  std::unique_ptr<MacroInvocation> macro;
+
+public:
+  AttrInputMacro (std::unique_ptr<MacroInvocation> macro)
+    : macro (std::move (macro))
+  {}
+
+  AttrInputMacro (const AttrInputMacro &oth)
+    : macro (oth.macro->clone_macro_invocation_impl ())
+  {}
+
+  AttrInputMacro (AttrInputMacro &&oth) : macro (std::move (oth.macro)) {}
+
+  void operator= (const AttrInputMacro &oth)
+  {
+    macro = oth.macro->clone_macro_invocation_impl ();
+  }
+
+  void operator= (AttrInputMacro &&oth) { macro = std::move (oth.macro); }
+
+  std::string as_string () const override
+  {
+    return " = " + macro->as_string ();
+  }
+
+  void accept_vis (ASTVisitor &vis) override;
+
+  // assuming this can't be a cfg predicate
+  bool check_cfg_predicate (const Session &) const override { return false; }
+
+  // assuming this is like AttrInputLiteral
+  bool is_meta_item () const override { return false; }
+
+  std::unique_ptr<MacroInvocation> &get_macro () { return macro; }
+
+  AttrInputType get_attr_input_type () const final override
+  {
+    return AttrInput::AttrInputType::MACRO;
+  }
+
+protected:
+  AttrInputMacro *clone_attr_input_impl () const override
+  {
+    return new AttrInputMacro (*this);
+  }
+};
```

Comment:
this is a nitpick, but it's true especially for new classes - it would be nice to have the definitions reside in a source file, so `rust-ast.cc` or create a new `rust-expr.cc` which would be nice too. This can be addressed in a separate PR and you can ignore this if you want.